proofs/proof_global.ml

```diff
@@ -372,52 +321,40 @@ let return_proof ?(allow_partial=false) () =
     List.map (fun (c, _) -> (proof_opt c, eff)) initial_goals in
     proofs, Evd.evar_universe_context evd
 
-let close_future_proof ~opaque ~feedback_id proof =
-  close_proof ~opaque ~keep_body_ucst_separate:true ~feedback_id ~now:false proof
-let close_proof ~opaque ~keep_body_ucst_separate fix_exn =
+let close_future_proof ~opaque ~feedback_id (ps, psl) proof =
+  close_proof ~opaque ~keep_body_ucst_separate:true ~feedback_id ~now:false proof ps
+
+let close_proof ~opaque ~keep_body_ucst_separate fix_exn (ps, psl) =
   close_proof ~opaque ~keep_body_ucst_separate ~now:true
-    (Future.from_val ~fix_exn (return_proof ()))
+    (Future.from_val ~fix_exn (return_proof (ps,psl))) ps
 
 (** Gets the current terminator without checking that the proof has
     been completed. Useful for the likes of [Admitted]. *)
-let get_terminator () = CEphemeron.get ( cur_pstate() ).terminator
-let set_terminator hook =
-  match !pstates with
-  | [] -> raise NoCurrentProof
-  | p :: ps -> pstates := { p with terminator = CEphemeron.create hook } :: ps
-
-module V82 = struct
-  let get_current_initial_conclusions () =
-  let { proof; strength } = cur_pstate () in
-  let Proof.{ name; entry } = Proof.data proof in
-  let initial = Proofview.initial_goals entry in
-  let goals = List.map (fun (o, c) -> c) initial in
-  name, (goals, strength)
-end
-
-let freeze ~marshallable =
-  if marshallable then CErrors.anomaly (Pp.str"full marshalling of proof state not supported.")
-  else !pstates
-let unfreeze s = pstates := s
-let proof_of_state = function { proof }::_ -> proof | _ -> raise NoCurrentProof
+let get_terminator (ps, _) = CEphemeron.get ps.terminator
+let set_terminator hook (ps, psl) =
+  { ps with terminator = CEphemeron.create hook }, psl
+
 let copy_terminators ~src ~tgt =
-  assert(List.length src = List.length tgt);
-  List.map2 (fun op p -> { p with terminator = op.terminator }) src tgt
+  let (ps, psl), (ts,tsl) = src, tgt in
+  assert(List.length psl = List.length tsl);
+  {ts with terminator = ps.terminator}, List.map2 (fun op p -> { p with terminator = op.terminator }) psl tsl
 
-let update_global_env pf_info =
+let update_global_env (pf : t) =
+  let res, () =
   with_current_proof (fun _ p ->
      Proof.in_proof p (fun sigma ->
        let tac = Proofview.Unsafe.tclEVARS (Evd.update_sigma_env sigma (Global.env ())) in
        let (p,(status,info)) = Proof.run_tactic (Global.env ()) tac p in
-         (p, ())))
+         (p, ()))) pf
+  in res
 
 (* XXX: Bullet hook, should be really moved elsewhere *)
-let () =
-  let hook n =
-    try
-      let prf = give_me_the_proof () in
-      (Proof_bullet.suggest prf)
-    with NoCurrentProof -> mt ()
-  in
-  Proofview.set_nosuchgoals_hook hook
+(* let _ =
```

Comment:
What's going on?

Reply:
This is a TODO indeed, what is this hook stuff doing? Nothing broke by removing it.

Reply:
Ok, I just commented the TODOs so that we don't miss them.

I'm not an expert, but I think the hook is for `Proof using` suggestions. I'm surprised we don't have a single test for it, though. @gares ?

Reply:
@gares ping, can this be removed or not?

Reply:
I don't think this is about `Proof using`.
It is about writing the wrong bullet, eg
```
elim...
  + bla...
     bla....
  - (* Error, you should use + here *)
```

So I'm not competent. Maybe @Zimmi48 knows how bullet works and what this hook is for.

Reply:
See #9646 for discussion on how to better implement this check.

Reply:
The proposed approach makes sense to me.
Given that it is a few lines I don't get why it is not a commit in this PR since:
- we got a discussion about a solved problem
- if we don't merge both we get a little regression in error messages

Reply:
The approach on that PR still needs some refinement, so it is not ready to merge as-is; the rationale for deferring to a separate PR is that it makes discussion possible; whereas here the PR is already too large and too delayed as to have useful discussion with the current review interface.

As you point out, topics have been re-discussed a few times, including addressing this concern when the PR was originally submitted.